fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/codec/LuceneOptimizedCompoundReader.java

```diff
@@ -0,0 +1,218 @@
+ * This class implements a directory, but is limited to only read operations.
+ * Directory methods that would normally modify data throw an exception.
+ * @lucene.experimental
+ */
+final class LuceneOptimizedCompoundReader extends Directory {
+
+    private final Directory directory;
+    private final String segmentName;
+    private final Map<String, FileEntry> entries;
+    private final IndexInput handle;
+
+    /** Offset/Length for a slice inside of a compound file. */
+    public static final class FileEntry {
+        long offset;
+        long length;
+    }
+
+    /**
+     * Create a new CompoundFileDirectory.
+     */
+    // TODO: we should just pre-strip "entries" and append segment name up-front like simpletext?
+    // this need not be a "general purpose" directory anymore (it only writes index files)
+    public LuceneOptimizedCompoundReader(Directory directory, SegmentInfo si, IOContext context) throws IOException {
+        this.directory = directory;
+        this.segmentName = si.name;
+        String entriesFileName = IndexFileNames.segmentFileName(segmentName, "", LuceneOptimizedCompoundFormat.ENTRIES_EXTENSION);
+        this.entries = readEntries(si.getId(), directory, entriesFileName);
+        String dataFileName = IndexFileNames.segmentFileName(segmentName, "", LuceneOptimizedCompoundFormat.DATA_EXTENSION);
+        handle = directory.openInput(dataFileName, context);
+        /* REMOVED TO NOT READ THE CHECKSUM ON EACH SEARCH
+        try {
+            CodecUtil.checkIndexHeader(handle, LuceneOptimizedCompoundFormat.DATA_CODEC, version, version, si.getId(), "");
+
+            // NOTE: data file is too costly to verify checksum against all the bytes on open,
+            // but for now we at least verify proper structure of the checksum footer: which looks
+            // for FOOTER_MAGIC + algorithmID. This is cheap and can detect some forms of corruption
+            // such as file truncation.
+            CodecUtil.retrieveChecksum(handle);
+
+            // We also validate length, because e.g. if you strip 16 bytes off the .cfs we otherwise
+            // would not detect it:
+            if (handle.length() != expectedLength) {
+                throw new CorruptIndexException("length should be " + expectedLength + " bytes, but is " + handle.length() + " instead", handle);
+            }
+
+            success = true;
+        } finally {
+            if (!success) {
+                IOUtils.closeWhileHandlingException(handle);
+            }
+        }
+         */
+    }
+
+    /** Helper method that reads CFS entries from an input stream. */
+    @SuppressWarnings("PMD.AvoidCatchingThrowable")
+    private Map<String, FileEntry> readEntries(byte[] segmentID, Directory dir, String entriesFileName) throws IOException {
+        Map<String, FileEntry> mapping = null;
+        try (ChecksumIndexInput entriesStream = dir.openChecksumInput(entriesFileName, IOContext.READONCE)) {
```

Comment:
Aren't we trying to avoid checksumming the files?

Reply:
We are trying to avoid reading the last block of each data file to retrieve the checksum.  I did not change this code because it is reading a byte[] serialized into the FileReference record.